stan/math/prim/prob/wiener4_lcdf.hpp

```diff
@@ -0,0 +1,863 @@
+template <typename T_x>
+inline auto rexp(T_x&& x) noexcept {
+  return (x <= 700) ? exp(x) : exp(700);
+}
+
+/**
+ * Helper function for Mill's ratio. This is dnorm on log-scale.
+ *
+ * @param x A scalar
+ * @return dnorm on log-scale
+ */
+template <typename T_x>
+inline auto lognormal(T_x&& x) noexcept {
+  return -0.5 * x * x - LOG_SQRT_PI - 0.5 * LOG_TWO;
+}
+
+/**
+ * Helper function for Mill's ratio. This is pnorm on log-scale.
+ *
+ * @param z A scalar
+ * @return pnorm on log-scale
+ *
+ * The original function lnnorm was written by ...
+ *     Jean Marie Linhart
+ *     StataCorp LP
+ *     jlinhart@stata.com
+ *     January 4, 2008
+ * and later modified and adapted to Stan
+ */
+template <typename T_z>
+inline auto lnnorm(T_z&& z) noexcept {
+  using ret_t = return_type_t<T_z>;
+  static const double LNNORM_MAX_X = 38.0;
+  static const double LNNORM_MIN_X = -1e9;
+  int lower;
+
+  auto t = ret_t(0.0);
+  auto z_ = z;
+  auto s = z_;
+
+  if (z_ == 0.0) {
+    return ret_t(log(0.5));
+  }
+  if (z_ > LNNORM_MAX_X) {
+    return ret_t(0.0);
+  }
+  if (z_ <= LNNORM_MIN_X) {
+    return ret_t(-0.5 * z_ * z_);
+  }
+  if (z_ < 0.0) {
+    z_ = -z_;
+    lower = 1;
+  } else {
+    lower = 0;
+  }
+
+  const auto z2 = square(z_);
+  const auto y = exp(-0.5 * z2) / SQRT_TWO_PI;
+  auto n = y / z_;
+
+  if (!(z_ > 2.0)) {
+    z_ *= y;
+    s = z_;
+    for (n = 3; s != t; n += 2) {
+      t = s;
+      z_ *= (z2 / n);
+      s += z_;
+    }
+    if (lower) {
+      return ret_t(log(0.5 - s));
+    }
+    return ret_t(log(0.5 + s));
+  }
+
+  auto a1 = ret_t(2.0);
+  auto a2 = ret_t(0.0);
+  n = z2 + 3;
+  auto p1 = ret_t(1.0);
+  auto q1 = z_;
+  auto p2 = (n - 1);
+  auto q2 = n * z_;
+  auto m = p1 / q1;
+  t = p2 / q2;
+  s = m;
+
+  for (n += 4; m != t && s != t; n += 4) {
+    a1 -= 8.0;
+    a2 += a1;
+    s = a2 * p1 + n * p2;
+    p1 = p2;
+    p2 = s;
+    s = a2 * q1 + n * q2;
+    q1 = q2;
+    q2 = s;
+    s = m;
+    m = t;
+    if (q2 > 1.0e30) {
+      p1 /= 1.0e30;
+      p2 /= 1.0e30;
+      q1 /= 1.0e30;
+      q2 /= 1.0e30;
+    }
+    t = p2 / q2;
+  }
+  t = lower ? log(t) - 0.5 * z2 - log(SQRT_TWO_PI) : log1p(-y * t);
+  return ret_t(t);
+}
+
+/**
+ * Helper function. Log of Mill's ratio for the normal distribution
+ *
+ * @param x A scalar
+ * @return Log of Mill's ratio
+ */
+template <typename T_x>
+inline auto logMill(T_x&& x) noexcept {
+  if (x > 1.0e5) {
+    return -log(x);
+  }
+  const auto m = lnnorm(-x) - lognormal(x);
+  return m;
+}
+
+/**
+ * Calculate the probability term 'P' on log scale
+ *
+ * @param a The boundary separation
+ * @param v The drift rate
+ * @param w The relative starting point
+ * @return 'P' term
+ */
+template <typename T_a, typename T_w, typename T_v>
+inline auto log_probability_distribution(const T_a& a, const T_v& v,
+                                         const T_w& w) noexcept {
+  using ret_t = return_type_t<T_a, T_w, T_v>;
+  auto nearly_one = ret_t(1.0 - 1.0e-6);
```

Comment:
@bob-carpenter this is another spot where I'm getting a bit confused. Here the code is checking if we are within 1.0e-6 of 1 and then branching off of that. How do we normally handle cases like this?